compiler/rustc_resolve/src/late/lifetimes.rs

```diff
@@ -2105,6 +2137,43 @@ fn is_late_bound_map(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<&FxIndexSet<
                     // (defined above)
                 }
 
+                hir::TyKind::Path(hir::QPath::Resolved(
+                    None,
+                    hir::Path { res: Res::Def(DefKind::TyAlias, alias_def), segments, .. },
+                )) => {
+                    // If this is a top level type alias attempt to "look through" it to see if the args
+                    // are constrained, instead of assuming they are and inserting all the lifetimes.
+                    // This is necessary for the following case:
+                    // ```
+                    // type Alias<'a, T> = <T as Trait<'a>>::Assoc;
+                    // fn foo<'a>(_: Alias<'a, ()>) -> Alias<'a, ()> { ... }
+                    // ```
+                    // If we considered `'a` constrained then it would become late bound causing an error
+                    // during astconv as the `'a` is not constrained by the input type `<() as Trait<'a>>::Assoc`
+                    // but appears in the output type `<() as Trait<'a>>::Assoc`.
+
+                    let generics = self.tcx.generics_of(alias_def);
+                    let mut walker = ConstrainedCollectorPostAstConv {
+                        arg_is_constrained: vec![false; generics.params.len()].into_boxed_slice(),
+                    };
+                    walker.visit_ty(self.tcx.type_of(alias_def));
+
+                    match segments.last() {
+                        Some(hir::PathSegment { args: Some(args), .. }) => {
+                            for (_, constrained_arg) in args
+                                .args
+                                .iter()
+                                .enumerate()
+                                .filter(|(n, _)| walker.arg_is_constrained[*n])
```

Comment:
You are using the index inside `args` to index a vector which is indexed by a `ty::Generics` index. Those may not match: matching args with params is a 130 line-long code snippet in `astconv::generics`.
However, I'm not sure we could even encounter a mismatch in well-formed code in the arguments to a `TyAlias` (we certainly can on a `Fn`).

Reply:
ah yeah it probably ought to be `.get(*n).unwrap_or_else(|| bug!())` :thinking: I did think that it _probably_ wouldn't be as simple as this logic but wasn't sure what it _should_ be. going to take a look at `astconv::generics`